### ISSUE TITLE
Monthly Performance Meetings: 2022

### ISSUE THREAD

**Issue by fcollonval** [jupyterlab/benchmarks#88]
Here is a record of the [agenda/minutes](https://hackmd.io/DTEJ5sy9Q9yeW0gYuX0v6A) for our monthly meeting. Everyone is welcome to join.

- When: Every 4 weeks Wednesday [10AM Pacific Time](https://www.thetimezoneconverter.com/?t=10%3A00%20am&tz=San%20Francisco&)
- Where: [`jovyan` Zoom](https://zoom.us/my/jovyan?pwd=c0JZTHlNdS9Sek9vdzR3aTJ4SzFTQT09)

Let's avoid using this thread for discussion. If you'd like to discuss something in the minutes, open a separate issue and reference this thread.

## Meeting Notes

- [26 Oct 2022](https://github.com/jupyterlab/benchmarks/issues/88#issuecomment-1293141415)
- [28 Sep 2022](https://github.com/jupyterlab/benchmarks/issues/88#issuecomment-1261899079)
- [3 Aug 2022](https://github.com/jupyterlab/benchmarks/issues/88#issuecomment-1204904103)
- [6 Jul 2022](https://github.com/jupyterlab/benchmarks/issues/88#issuecomment-1177348917)
- [8 Jun 2022](https://github.com/jupyterlab/benchmarks/issues/88#issuecomment-1150720591)
- [12 May 2022](https://github.com/jupyterlab/benchmarks/issues/88#issuecomment-1124581801)
- [13 Apr 2022](https://github.com/jupyterlab/benchmarks/issues/88#issuecomment-1098307788)
- [16 Mar 2022](https://github.com/jupyterlab/benchmarks/issues/88#issuecomment-1070385376)
- [16 Feb 2022](https://github.com/jupyterlab/benchmarks/issues/88#issuecomment-1042653217)
- [19 Jan 2022](https://github.com/jupyterlab/benchmarks/issues/88#issuecomment-1016753482)

**Comment by fcollonval** [comment]
### 19 Jan 2022 Monthly meeting

#### Attendees

| Name     | affiliation  | github username  |
| -------- | --------------- | --------|
| Frederic Collonval | QuantStack | @fcollonval |
| William Stein | CoCalc/SageMath | @williamstein|
| Marc Udoff | DeShaw | | 
| Paul Ivanov | Noteable | @ivanov |

#### Notes

- Marc
    - Limit outputs work done  
    It will probably become open-source in the near future
    Make long notebook more manageable
    
- Frederic
    - Search extension: [PR - Rework to carry search on the model](https://github.com/jupyterlab/jupyterlab/pull/11689)
        - Document decision done: like ignoring the link markdown
        - Keep the GenericSearch for not supported outputs
        - william: encourage you to document the choices related to the subtle things like links, math in markdown, ansi output messages.  I would be thrilled if there was an official jupyter specification of what search should do, and then client implementors could try to follow that spec.
        - Output messages is by far the hardest part, due to lack of control over them.
        - Regular expressions supported?  Yes.
        - Look at other tools like VSCode to see what they are doing

**Comment by fcollonval** [comment]
### 16 Feb 2022 Monthly meeting

#### Attendees

| Name     | affiliation  | github username  |
| -------- | --------------- | --------|
| Eric Charles |  | @echarles |
| Frederic Collonval | QuantStack | @fcollonval |
|  William Stein| SageMath/CoCalc | @williamstein |

#### Notes

- echarles
  - https://github.com/datalayer/jupyter-react/pull/10

  - Tools for memory leak tracking
    - from gitter by Mike
    > This looks interesting for debugging memory leaks: https://blogs.windows.com/msedgedev/2021/12/09/debug-memory-leaks-detached-elements-tool-devtools/
    - [nick] more API coming in webdriver to get access to the memory allocation: https://gist.github.com/nolanlawson/01e7358f682dde01e9cb27f777acf195  
  Done some work on JupyterLite

  - Search feature: will work on making the search debouncer customizable and to add triggering the search on `enter`.
  
    Looking at what VS Code is doing: the search up to 1.63 was scoped to the notebook code source (and not showing in rendered markdown cell - the cell state was switched when the current search match was in a markdown cell). But since 1.64, they display search in the rendered markdown cells and they have an option to search in outputs. But they warn about performance issue, because in that case they will rendered the outputs prior to search in it.

  - Web workers
  Some resources from Nick:  
  https://developers.google.com/web/tools/workbox 
  https://github.com/GoogleChromeLabs/comlink/ 
  https://github.com/jupyterlite/jupyterlite/pull/137 

- Frederic
    - Browser hangs for notebook with hundreds of MB in metadata property  
    https://github.com/jupyterlab/benchmarks/issues/89

    > Should the user issue be stored in the the benchmark repo or the core one

    - Starting debugger is slow  
    https://github.com/jupyterlab/jupyterlab/issues/12012
    
    - Scheduled benchmark tests
    After investigation no degredation was found. Seems we got unlucky with the CI machine allocated.  
    Needs to split better the test cases

    - Back to performance work next week
        1. Finish search from model [PR](https://github.com/jupyterlab/jupyterlab/pull/11689)
        2. Update benchmark repo with those new cases (doc and CI)
        3. Rework ToC extension to be based on data model
        4. Virtual notebook
    

- William: vague remark...
   - I'm implementing a "collaborative whiteboard" version of Jupyter notebooks right now, which leads to interesting performance questions. 
   - https://github.com/sagemathinc/cocalc/pull/5674#issuecomment-1040570390

- Memory Leaks - maybe testing for in an automated way:
   - https://gist.github.com/nolanlawson/01e7358f682dde01e9cb27f777acf195

- Nick Bollweg: Making webworkers more usable, e.g., for search
   - https://github.com/GoogleChromeLabs/comlink/
   - https://developers.google.com/web/tools/workbox
   - https://github.com/jupyterlite/jupyterlite/pull/137
   - comments on subtle performance implications of using react...

**Comment by fcollonval** [comment]
### 16 March 2022

#### Attendees

| Name     | affiliation  | github username  |
| -------- | --------------- | --------|
| Frederic Collonval | QuantStack | @fcollonval |
| Eric Charles | Datalayer | @echarles |
| Marc Udoff | D. E. Shaw | @mlucool |
 
#### Notes

- Frederic
    - Final version of [search from model](https://github.com/jupyterlab/jupyterlab/pull/11689)  
    At the end the search on output is still on the HTML. To hard to figure out what will be in the view (after testing mainly on Markdown).
    
- Marc
    - ad blocker hit performance in Chrome really bad - we should document and shout out that information
    - Works by rtrg: 
        - [CSS containment](https://web.dev/content-visibility/) discusstion in [that comment](https://github.com/jupyterlab/jupyterlab/pull/11074#issuecomment-1065150039) - unfortunately not available in FireFox.  
        Context: clicking on the context menu

    ```
    %%javascript
    function addCss(cssCode) {
        const style = document.createElement('style');
        style.innerHTML = cssCode;
        document.head.appendChild(style);
    }

    addCss(`
        .jp-NotebookPanel.jp-Document {
            will-change: transform;
            transform: none !important;
        }
        .jp-NotebookPanel.jp-Document.lm-Widget.lm-mod-hidden {
            transform: scaleX(0) !important;
            display: block !important;
        }
    `);

    // This uses experimental CSS that is not supported by all browsers.
    // By setting content-visibility: auto, when the cell is off-screen,
    // style checks are avoided. We set an intrinsic height so that
    // scrolling is not too jittery.
    addCss(`
        .jp-Cell {
            content-visibility: auto !important;
            contain-intrinsic-height: auto 70px;
        }
    `);
    ```

    >  The above shows great improvements.

    - Try hidding the panels with CSS containment for speeding-up tab switching. It makes rendering of new tabs slower but then switch faster.
    - [jupyterlab-limit-output](https://github.com/deshaw/jupyterlab-limit-output) can help user for with those long outputs  
    User can blow their browser by memory size. There may be a memory leak somewhere - but enable to isolate it for now.
    - Loading static assets [Jupyter Server issue](https://github.com/jupyter-server/jupyter_server/issues/312)
    In Cocalc the assets are loaded using NGinx

- Eric
    - Work on the kernel: On kernel area, recent work on kernel message parsing is merged https://github.com/jupyter-server/jupyter_server/pull/657 
    - Virtual deferred rendering: 
        - renderCellOnIdle improvements by waiting for _long enough_ idle time [#12166](https://github.com/jupyterlab/jupyterlab/pull/12166)
        - Working on a PR to block cell list edition when the notebook is not yet fully rendered.
        - Probably gonna revert the setting to not activate virtual rendering by default

**Comment by fcollonval** [comment]
### 16 April 2022

#### Attendees

| Name     | affiliation  | github username  |
| -------- | --------------- | --------|
| Frederic Collonval | QuantStack | @fcollonval |
| Marc Udoff | D. E. Shaw & Co. |  @mlucool |
| William Stein | CoCalc | @williamstein |
 
#### Notes

- Frederic
    - Windowing notebooks:
        - [Table of content refactor PR](https://github.com/jupyterlab/jupyterlab/pull/12374)
        - Next step windowing notebooks
    - React is working on reloading state of unmounted components for case like virtual rendering or tab switch: https://github.com/reactwg/react-18/discussions/19
    - CSS containment landed in Webkit 15.4: https://webkit.org/blog/12445/new-webkit-features-in-safari-15-4/#css

- William
    - I'm back to windowing for CoCalc because of examples/complaints from users, including this math-heavy example: https://nbviewer.org/url/archive.org/download/musing_math/Volume_Solutions_To_Snub_Dodecahedron.ipynb
        - Good be interesting to add a benchmark test on that one.
        - Using [React Windowing](https://github.com/bvaughn/react-window) and move to
    - https://virtuoso.dev/
    - https://github.com/sagemathinc/cocalc/pull/
    - planning to render iframes differently to handle: https://github.com/sagemathinc/cocalc/issues/4727
    - Shared issues with state storage with Plotly charts and windowing library.
       - [example in pyinstrument](https://github.com/joerick/pyinstrument/blob/main/pyinstrument/magic/magic.py#L103); works by shoving output into an iframe.

- Marc
    - Firefox vs. Chrome: major improvements for the user experience in Firefox; mainly linked to the hit tests in Chrome. See 2 order of magnitude improvements.

**Comment by fcollonval** [comment]
### 11 May 2022

#### Attendees

| Name     | affiliation  | github username  |
| -------- | --------------- | --------|
| Jason Grout | Databricks | @jasongrout |
| WilliamStein | CoCalc | @williamstein|
| Marc Udoff | D. E. Shaw & Co | @mlucool |
| Pete Blois | Google | |
| Frederic Collonval | QuantStack | @fcollonval |


#### Notes

- Frederic
    - Windowing notebooks [PR](https://github.com/jupyterlab/jupyterlab/pull/12554)
        - To deal with outputs internal state, the current envision solution is to keep the outputs in the DOM.
        - iframe does not support to being moved in the DOM tree.

- William
    - Windowing using [Virtuoso](https://virtuoso.dev/) is live in cocalc, and also used everywhere in cocalc that involves lists.
    - Solved the iframe state problem  
      The solution used is to place those elements in the header element (Virtuoso feature) and move it at the right position
      when the cell is in the view port.
    - Also solved for ipywidgets - k3d at least.
    - [Pete] Do you deal with jumping to link within the document? It is not supported in Cocalc. (but should be...)

**Comment by fcollonval** [comment]
### 8 June 2022

#### Attendees

| Name     | affiliation  | github username  |
| -------- | --------------- | --------|
| Frederic Collonval | QuantStack | @fcollonval |
| Jason Grout | Databricks| @jasongrout |
|William Stein | CoCalc/SageMath | @williamstein |



#### Notes

- Virtualization in CoCalc:
  - I finally got one user complaint about my windowing implementation.  %%html blocks with styles... the style isn't global and disappears when the style block leaves the DOM.
  - Solution: scan the cells to move those in the header hidden cell to be available globally.
    - Currently styles are accumulated, and not updated as the html block changes. This does seem to satisfy users, though.
    - But, that could change, actually.   Just an issue of time.
    - Other comment: i didn't worry about style blocks in markdown html... yet.

- Frederic
    - Windowing notebooks [PR](https://github.com/jupyterlab/jupyterlab/pull/12554)  
    ![](https://i.imgur.com/N9ZJnNY.png)
    
    Drag-and-drop with copy: we should not detach the source cell
    What about the case of %%html
    What about RTC structure change like added cell
    Listing to DOM events to see if the iframe is pulled out when scrolling into a specific position
    
    - william UX question: how do you enable/disable windowing; thoughts about *scope* of this?  
    Currently there is a single user settings for all documents.
    But it is propagated to all notebook documents. So it will be
    easy to set it per document (same pattern as line numbers e.g.).
    About storing it within the document (aka notebook metadata);
    I did not thought about it and I am a bit reluctant to store
    view parameters in the document model.
    [william] The windowing parameter is usually conditioned by
    the document content (e.g. the type of outputs used). So it
    would make sense to store it in the document.
    - william question 2: any relations/issues between RTC and your virtualization implimentation?  
    It has not been tested in RTC context.
    - william: I didn't know about "scroll to input"...

    - Links for the browser supporting virtualization:
      -  https://github.com/WICG/display-locking/
      -  [< virtual-scroller >](https://chromestatus.com/feature/5673195159945216): chrome no longer pursuing
      -  [Aria virtual content](https://chromestatus.com/feature/5665065770418176): Chrome no active development
      - [Custom highlighting](https://chromestatus.com/feature/5436441440026624): chrome proposing
    - Memory leak benchmark  
    Start using [fuite](https://github.com/nolanlawson/fuite)  
    Trying [BLeak](https://github.com/plasma-umass/BLeak) - failed to make it works

**Comment by fcollonval** [comment]
### 6 July 2022

#### Attendees

| Name     | affiliation  | github username  |
| -------- | --------------- | --------|
| Frederic Collonval | QuantStack | @fcollonval |
|William Stein | CoCalc/SageMath | @williamstein |

#### Notes

- Frederic
    - Performance:
        - CodeMirror 6 alone will bring a great boost: see [benchmark](https://github.com/jupyterlab/jupyterlab/pull/11638#issuecomment-1152587974)
    - Memory leaks:
        - CI job at https://github.com/jupyterlab/benchmarks/pull/100 using [fuite](https://github.com/nolanlawson/fuite)  
        _fuite_ outputs does not always matches results obtained with the [3 snapshots technique](http://slides.com/gruizdevilla/memory#/7) for the leaking objects and the memory delta may be the opposite - I trusted more the objects and collections leak information than the memory size.
        - PRs:
            - [Clear `Title.changed` signal](https://github.com/jupyterlab/lumino/pull/308)
            - JupyterLab PRs: [12750](https://github.com/jupyterlab/jupyterlab/pull/12750), [12774](https://github.com/jupyterlab/jupyterlab/pull/12774) and [12777](https://github.com/jupyterlab/jupyterlab/pull/12777)  
              Trying to get them for the next JupyterLab 3.4.x patch release.


Discussion of how to deal with memory leaks happened.

What cocalc does is this.  But WeakRef's can be even better...

```ts
export function close(obj: object, omit?: Set<string>): void {
  if (omit != null) {
    Object.keys(obj).forEach(function (key) {
      if (omit.has(key)) return;
      if (typeof obj[key] == "function") return;
      delete obj[key];
    });
  } else {
    Object.keys(obj).forEach(function (key) {
      if (typeof obj[key] == "function") return;
      delete obj[key];
    });
  }
}

```

This is similar to the disposable interface in Lumino. But with a more radical approach as it deletes all keys.

**Comment by fcollonval** [comment]
### 3 August 2022

#### Attendees

| Name     | affiliation  | github username  |
| -------- | --------------- | --------|
| Frederic Collonval | QuantStack | @fcollonval |
| A. T. Darian  | QuantStack | @afshin |
| M Krassowski |  | @krassowski |

#### Notes

- Darian
    - Lumino 2: Try to test passive event but it breaks the test.  
    https://chromestatus.com/feature/5745543795965952  
    Probably should be applied only to some specific actions.  
    - https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener  
    If not specified, defaults to false – except that in browsers other than Safari and Internet Explorer, defaults to true for the wheel, mousewheel, touchstart and touchmove events. 
    - https://github.com/WICG/EventListenerOptions/blob/gh-pages/explainer.md

**Comment by fcollonval** [comment]
### 28 September 2022

#### Attendees

| Name     | affiliation  | github username  |
| -------- | --------------- | --------|
| Frederic Collonval | QuantStack | @fcollonval |
| M Krassowski |  | @krassowski |
|William Stein | CoCalc/SageMath | @williamstein |

#### Notes

- Frederic
    - Windowing notebooks [PR](https://github.com/jupyterlab/jupyterlab/pull/12554) is merged  
    It can be tested on the pre release `pip install --pre jupyterlab`
    - Revert using transform scale for switching tabs [JupyterLab PR](https://github.com/jupyterlab/jupyterlab/pull/13103)  
    It was crashing the UI with tabs containing iframe (such as documentation tab) in Chrome-based browser. So this is now adviced to be used only on Firefox (not tested on webkit)
    To be released in 3.4.8
- Mike
  - looking at menu performance

**Comment by fcollonval** [comment]
### 26 October 2022

#### Attendees

| Name     | affiliation  | github username  |
| -------- | --------------- | --------|
| Frederic Collonval | QuantStack | @fcollonval |
| A. T. Darian | QuantStack | @afshin |
| M Krassowski | Quansight | @krassowski |
|William Stein | CoCalc/SageMath | @williamstein |
| Vidar T Fauske | JP Morgan Chase | @vidartf |

#### Notes

- [Mike] UI benchmark tool: [PR](https://github.com/jupyterlab/benchmarks/pull/122)
- [Frederic] Benchmarks project is currently broken due to unsupported syntax (nullish assignment operator) in vega-lite by older typescript version
- [Darian] Note about `MessageLoop` internal `schedule` implementation in Lumino 2  
  In Lumino 2, the messages are gonna be processed in micro-tasks (no more in macro-tasks like in Lumino 1 due to `setImmediate` or `requestAnimationFrame` usage). So in pathologic case, this can create infinite loop (e.g. if an widget update method calls itself recursively).
- [Vidar] Suggestion for adding more realistic benchmarking case with multiple large files open: https://github.com/jupyterlab/benchmarks/issues/124
- [Mike]
  - Delayed CSS style loading
  - MathJax 4/[MathML support in browsers](https://chromestatus.com/feature/5240822173794304)
    - do we need server URL for MathJax
